src/H5T.c

```diff
@@ -5157,80 +5223,40 @@ H5T__path_find_real(const H5T_t *src, const H5T_t *dst, const char *name, H5T_co
-        H5T_g.path[0]->conv.u.lib_func = H5T__conv_noop;
-        H5T_g.path[0]->cdata.command   = H5T_CONV_INIT;
-        if (H5T__conv_noop(NULL, NULL, &(H5T_g.path[0]->cdata), &tmp_ctx, 0, 0, 0, NULL, NULL) < 0) {
-#ifdef H5T_DEBUG
-            if (H5DEBUG(T))
-                fprintf(H5DEBUG(T), "H5T: unable to initialize no-op conversion function (ignored)\n");
-#endif
-            /* Ignore any errors from the conversion function */
-            if (H5E_clear_stack(NULL) < 0)
-                HGOTO_ERROR(H5E_DATATYPE, H5E_CANTRESET, NULL, "unable to clear current error stack");
-        } /* end if */
-        H5T_g.path[0]->is_noop = true;
-        H5T_g.npaths           = 1;
-    } /* end if */
+    /* Make sure the path table is initialized */
+    if ((0 == H5T_g.npaths) && (H5T__path_find_init_path_table() < 0))
+        HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "unable to initialize type conversion path table");
 
-    /* Find the conversion path.  If source and destination types are equal
-     * then use entry[0], otherwise do a binary search over the
-     * remaining entries.
-     *
-     * Only allow the no-op conversion to occur if no "force conversion" flags
-     * are set
+    /* Find the conversion path. If no "force conversion" flags are
+     * set and the source and destination types are equal, then use
+     * the no-op conversion path. Otherwise, do a binary search over
+     * the remaining entries.
      */
-    if (src->shared->force_conv == false && dst->shared->force_conv == false &&
-        0 == H5T_cmp(src, dst, true)) {
-        table = H5T_g.path[0];
-        cmp   = 0;
-        md    = 0;
-    } /* end if */
-    else {
-        lt = md = 1;
-        rt      = H5T_g.npaths;
-        cmp     = -1;
-
-        while (cmp && lt < rt) {
-            md = (lt + rt) / 2;
-            assert(H5T_g.path[md]);
-            cmp = H5T_cmp(src, H5T_g.path[md]->src, false);
-            if (0 == cmp)
-                cmp = H5T_cmp(dst, H5T_g.path[md]->dst, false);
-            if (cmp < 0)
-                rt = md;
-            else if (cmp > 0)
-                lt = md + 1;
-            else
-                table = H5T_g.path[md];
-        } /* end while */
-    }     /* end else */
+    noop_conv =
+        src->shared->force_conv == false && dst->shared->force_conv == false && 0 == H5T_cmp(src, dst, true);
+    if (noop_conv)
+        matched_path = H5T_g.path[0];
+    else
+        matched_path = H5T__bsearch_path_table(src, dst, &last_cmp, &path_idx);
 
     /* Keep a record of the number of paths in the table, in case one of the
      * initialization calls below (hard or soft) causes more entries to be
      * added to the table - QAK, 1/26/02
      */
     old_npaths = H5T_g.npaths;
 
+    /* Set a few convenience variables */
+    new_api_hard_func = (matched_path && conv->is_app && conv->u.app_func);
+    new_lib_hard_func = (matched_path && !conv->is_app && conv->u.lib_func);
```

Comment:
Suggest using 'new_lib_func' for a variable name, since it's not guaranteed to be a 'hard' conversion.    (Else, we wouldn't need the '!matched->is_hard' test, below)

Reply:
In the current implementation I think it is, in that `H5T__path_find_real` is currently only called in two places in the library: from the `H5T_path_find` function, in which case `app_func/lib_func` is always NULL, and from `H5T__register`, in which case `H5T__path_find_real` is only called when the passed in function is specified as `H5T_PERS_HARD`. That said, seems this could change in the future, so no problem renaming these to just "new_xxx_func".